### ISSUE TITLE
New naming schema for Amazon Linux 2 releases

### ISSUE THREAD

**Issue by artyomtkachenko** [inspec/inspec#3614]
### 🎛 Description

🐛 bug report

New naming schema for Amazon Linux 2 releases

New naming schema
```bash
cat /etc/system-release
Amazon Linux release 2 (Karoo)
```
Old naming scema
```bash
cat /etc/system-release
Amazon Linux 2
```
### 🌍 InSpec and Platform Version

Amazon Linux 2 
inspec version
3.0.46

### 🤔 Replication Case

```bash
docker run --rm -it  amazonlinux:2

cat /etc/system-release
Amazon Linux 2
bash-4.2# yum update -y
....

cat /etc/system-release
Amazon Linux release 2 (Karoo)
```

### 💁 Possible Solutions

train gem must be updated to address a new release name

### 💻 Stacktrace

```bash
Traceback (most recent call last):
	38: from /usr/bin/inspec:106:in `<main>'
	37: from /usr/bin/inspec:106:in `load'
	36: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/inspec-3.0.46/bin/inspec:12:in `<top (required)>'
	35: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/thor-0.20.0/lib/thor/base.rb:466:in `start'
	34: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/thor-0.20.0/lib/thor.rb:387:in `dispatch'
	33: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/thor-0.20.0/lib/thor/invocation.rb:126:in `invoke_command'
	32: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/thor-0.20.0/lib/thor/command.rb:27:in `run'
	31: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/inspec-3.0.46/lib/inspec/cli.rb:242:in `exec'
	30: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/inspec-3.0.46/lib/inspec/cli.rb:242:in `new'
	29: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/inspec-3.0.46/lib/inspec/runner.rb:58:in `initialize'
	28: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/inspec-3.0.46/lib/inspec/runner.rb:66:in `configure_transport'
	27: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/inspec-3.0.46/lib/inspec/backend.rb:51:in `create'
	26: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/transports/local.rb:17:in `connection'
	25: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/transports/local.rb:17:in `new'
	24: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/transports/local.rb:27:in `initialize'
	23: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/transports/local.rb:46:in `select_runner'
	22: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/plugins/base_connection.rb:114:in `platform'
	21: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect.rb:9:in `scan'
	20: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:27:in `scan'
	19: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:27:in `each'
	18: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:33:in `block in scan'
	17: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:45:in `scan_children'
	16: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:45:in `each'
	15: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:51:in `block in scan_children'
	14: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:60:in `scan_family_children'
	13: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:45:in `scan_children'
	12: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:45:in `each'
	11: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:51:in `block in scan_children'
	10: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:60:in `scan_family_children'
	 9: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:45:in `scan_children'
	 8: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:45:in `each'
	 7: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:51:in `block in scan_children'
	 6: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:60:in `scan_family_children'
	 5: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:45:in `scan_children'
	 4: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:45:in `each'
	 3: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:46:in `block in scan_children'
	 2: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:46:in `instance_eval'
	 1: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/specifications/os.rb:194:in `block in load'
/opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/helpers/os_linux.rb:17:in `redhatish_version': undefined method `[]' for nil:NilClass (NoMethodError)
```

**Comment by KasteM34** [comment]
Quick fix for me : 
change "Amazon Linux release 2 (Karoo)" to "Amazon Linux 2" in /etc/system-release.
And it works again then.

**Comment by jerryaldrichiii** [comment]
When this is merged ^ the problem should be fixed :smile:

**Comment by fatbasstard** [comment]
Currently Train is pinned: https://github.com/inspec/inspec/pull/3568

Can that be removed? 🤔

**Comment by jerryaldrichiii** [comment]
Closing since https://github.com/inspec/train/pull/380 is merged.

**Comment by nexeck** [comment]
It seems, that the dependency to the train package in inspec wasn't update?

With inspec-3.0.60-1.el7.x86_64 i keep getting the error.
It keeps using train version 1.5.6

```
/opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/helpers/os_linux.rb:16:in `redhatish_version': undefined method `[]' for nil:NilClass (NoMethodError)
```

**Comment by klub** [comment]
> It seems, that the dependency to the train package in inspec wasn't update?
> 
> With inspec-3.0.60-1.el7.x86_64 i keep getting the error.
> It keeps using train version 1.5.6
> 
> ```
> /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/helpers/os_linux.rb:16:in `redhatish_version': undefined method `[]' for nil:NilClass (NoMethodError)
> ```

Hi guys. Do you have any estimate when then train library that comes with inspec will be updated to contain a version that supports Amazon Linux 2 please? Are there any suggested work arounds you can advise in the meantime?

**Comment by nexeck** [comment]
@klub I am using this workaround:

```
# Fix until inspec gets update and uses train 1.5.10 (https://github.com/inspec/train/pull/380)
echo "Amazon Linux 2" > /etc/system-release

inspec detect
inspec exec /tmp/spec

**Comment by jerryaldrichiii** [comment]
@klub The release yesterday should have the newest version.

Can you update and try again?

**Comment by akr257** [comment]
I have the same error on an Amazon Linux2 instance

>     amazon-ebs: Installing inspec
>     amazon-ebs: installing with rpm...
>     amazon-ebs: warning: /tmp/install.sh.3535/inspec-3.0.64-1.el7.x86_64.rpm: Header V4 DSA/SHA1 Signature, key ID 83ef826a: NOKEY
>     amazon-ebs: Preparing...                          ########################################
>     amazon-ebs: You're about to install InSpec!
>     amazon-ebs: Updating / installing...
>     amazon-ebs: inspec-3.0.64-1.el7                   ########################################
>     amazon-ebs: Thank you for installing InSpec!
>     amazon-ebs: /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/helpers/os_linux.rb:16:in `redhatish_version': undefined method `[]' for nil:NilClass (NoMethodError)

**Comment by jaredledvina** [comment]
https://github.com/inspec/inspec/pull/3657 should bump `train` to the latest release which actually has the fix. If I understand correctly, that needs to be merged and another Inpsec release has to be cut.

**Comment by klub** [comment]
> @klub The release yesterday should have the newest version.
> 
> Can you update and try again?

Unfortunately its still failing with the same error:

```
5: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:45:in `scan_children'
	 4: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:45:in `each'
	 3: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:46:in `block in scan_children'
	 2: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/scanner.rb:46:in `instance_eval'
	 1: from /opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/specifications/os.rb:194:in `block in load'
/opt/inspec/embedded/lib/ruby/gems/2.5.0/gems/train-1.5.6/lib/train/platforms/detect/helpers/os_linux.rb:16:in `redhatish_version': undefined method `[]' for nil:NilClass (NoMethodError)
```

**Comment by dud225** [comment]
I can confirm that the bug is still present on train v1.6.3.
@artyomtkachenko could you please reopen this issue ?

**Comment by jerryaldrichiii** [comment]
@dud225 can you add the output of your stacktrace/error here?

**Comment by jerryaldrichiii** [comment]
I cannot reproduce this:

```[~]$ inspec -v
3.2.6
[~]$ inspec shell -t ssh://ec2-user@18.220.173.150 -i ~/.ssh/REDACTED.pem
Welcome to the interactive InSpec Shell
To find out how to use it, type: help

You are currently running on:

    Name:      amazon
    Families:  redhat, linux, unix, os
    Release:   2
    Arch:      x86_64

inspec>```

**Comment by jerryaldrichiii** [comment]
I cannot reproduce this with InSpec 3.2.6:

```
[~]$ inspec -v
3.2.6
[~]$ inspec shell -t ssh://ec2-user@18.220.173.150 -i ~/.ssh/REDACTED.pem

Welcome to the interactive InSpec Shell
To find out how to use it, type: help

You are currently running on:

    Name:      amazon
    Families:  redhat, linux, unix, os
    Release:   2
    Arch:      x86_64

inspec>
```

**Comment by dud225** [comment]
Hello @jerryaldrichiii 

I'm sorry I was wrong, I can now definitely confirm that the issue is indeed fixed.

I was not using Inspec directly but was relying on Kitchen-dokken. As it happens, Kitchen-dokken is using the local Inspec installation despites Inspec being also already present on the provisionner Docker image. So yesterday I just updated the Docker image, but left my local installation as is.

I've updated my local installation now and I'm no longer seeing the issue.
Thanks for your time and sorry again.
